service/history/conflictResolverV2.go

```diff
@@ -0,0 +1,191 @@
+	logger log.Logger) *conflictResolverImpl {
+
+	return &conflictResolverImpl{
+		shard:           shard,
+		clusterMetadata: shard.GetService().GetClusterMetadata(),
+		context:         context,
+		historyMgr:      historyMgr,
+		historyV2Mgr:    historyV2Mgr,
+		logger:          logger,
+	}
+}
+
+func (r *conflictResolverV2Impl) reset(prevRunID string, requestID string, replayEventID int64, info *persistence.WorkflowExecutionInfo, localVersionHistories persistence.VersionHistories) (mutableState, error) {
+	domainID := r.context.getDomainID()
+	execution := *r.context.getExecution()
+	startTime := info.StartTimestamp
+	eventStoreVersion := info.EventStoreVersion
+	branchToken := info.GetCurrentBranch()
+	replayNextEventID := replayEventID + 1
+
+	var resetMutableStateBuilder *mutableStateBuilder
+	var sBuilder stateBuilder
+	var lastEvent *shared.HistoryEvent
+	var lastFirstEventID int64
+
+	eventsToApply := replayNextEventID - common.FirstEventID
+
+	iter := collection.NewPagingIterator(func(paginationToken []byte) ([]interface{}, []byte, error) {
+		var paginateItems []interface{}
+		historyBatch, _, token, err := r.getHistoryBatch(domainID, execution, common.FirstEventID, replayNextEventID, paginationToken, eventStoreVersion, branchToken)
+		for _, history := range historyBatch {
+			paginateItems = append(paginateItems, history)
+		}
+		return paginateItems, token, err
+	})
+
+	for iter.HasNext() {
+		item, err := iter.Next()
+		if err != nil {
+			r.logger.Error("Conflict resolution err getting history.", tag.Error(err))
+			return nil, err
+		}
+
+		history := item.(*shared.History).Events
+		batchSize := int64(len(history))
+		// NextEventID could be in the middle of the batch.  Trim the history events to not have more events then what
+		// need to be applied
+		if batchSize > eventsToApply {
+			history = history[0:eventsToApply]
+		}
+		eventsToApply -= int64(len(history))
+
+		if len(history) == 0 {
+			break
+		}
+
+		firstEvent := history[0]
+		lastEvent = history[len(history)-1]
+		if firstEvent.GetEventId() == common.FirstEventID {
+			resetMutableStateBuilder = newMutableStateBuilderWithReplicationState(
+				r.clusterMetadata.GetCurrentClusterName(),
+				r.shard,
+				r.shard.GetEventsCache(),
+				r.logger,
+				firstEvent.GetVersion(),
+			)
+
+			resetMutableStateBuilder.executionInfo.EventStoreVersion = eventStoreVersion
+			sBuilder = newStateBuilder(r.shard, resetMutableStateBuilder, r.logger)
+		}
+
+		// NOTE: passing 0 as newRunEventStoreVersion is safe here, since we don't need the newMutableState of the new run
+		_, _, _, err = sBuilder.applyEvents(domainID, requestID, execution, history, nil, resetMutableStateBuilder.GetEventStoreVersion(), 0)
+		if err != nil {
+			r.logger.Error("conflict resolution err applying events.", tag.Error(err))
+			return nil, err
+		}
+		resetMutableStateBuilder.executionInfo.SetLastFirstEventID(lastFirstEventID)
```

Comment:
as long as state builder is applying history event by batch, the SetLastFirstEventID, SetNextEventID can be omitted (done by state builder)

Reply:
same applies to UpdateReplicationStateLastEventID